confmap/provider/templateprovider/provider.go

```diff
@@ -0,0 +1,109 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package templateprovider // import "go.opentelemetry.io/collector/confmap/provider/templateprovider"
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/confmap/provider/internal"
+)
+
+const (
+	schemeName  = "template"
+	templateKey = "template"
+	typeKey     = "type"
+
+	// The template provider will always return the following structure:
+	// {
+	//   "templates": {
+	//     "the_type_of_template": "the_template"
+	//   },
+	// }
+	// This allows multiple templates to be aggreated under the same global key.
+	allTemplatesKey = "templates"
+)
+
+type provider struct{}
+
+// New returns a new confmap.Provider that reads the template from a file.
+//
+// This Provider supports "file" scheme, and can be called with a "uri" that follows:
```

Comment:
Is there any way to take advantage of the other providers to handle reading the file contents? Since templates feel more like another type of config than a config "source" (file, http, s3, etc.), it would be unfortunate if templates could either only be obtained from select sources or that getting them from additional sources would require updates to the provider.

Reply:
Thanks for taking a look at this @evan-bradley. I worked out an implementation based only on a converter. (See https://github.com/open-telemetry/opentelemetry-collector/pull/8507) 

I actually like this a lot better.
- It has the advantage you mentioned, of allowing any provider to provide templates.
- It removes the awkward coupling between a provider and converter.
- It allows `templates` to appear directly in a config file (though I think typically they would not be used this way.) Still, it is nice to be able to reason about them in a more natural way.